src/internal/__tests__/curry.test.js

```diff
@@ -41,7 +41,52 @@ describe('variadicCurryWithValidation', function () {
     const f = variadicCurryWithValidation((x) => x === null, 'null', (x) => 0, func, false)
     expect(() => f(1)(2)(3)).toThrowError(new Error('func takes up to 2 arguments, followed by null. You already passed 3 arguments and the last argument was not null'))
   })
-  it('works on functions which take a variable number of iterables', function () {
 
+  describe('explicit partial application', function () {
+    let curriedMethod
+
+    beforeEach(function () {
+      curriedMethod = variadicCurryWithValidation((x) => '🙂🙁'.includes(x), 'emoji', emoji => '! ' + emoji, (a = '', b = 'goodbye', c) => a + b + c, false, 0, 2)
+    })
+
+    it('curried methods have a partial method', function () {
+      expect(typeof curriedMethod.partial).toBe('function')
+    })
+
+    it('works like calling the function if not all arguments have been applied', function () {
+      expect(curriedMethod.partial('hello ').partial('world')('🙂')).toBe('hello world! 🙂')
+    })
+
+    it('does nothing if there are no arguments', function () {
+      expect(curriedMethod.partial().partial('hello ').partial('world')('🙂')).toBe('hello world! 🙂')
+    })
+
+    it('can be used to defer application of the data parameter', function () {
+      expect(curriedMethod.partial().partial('hello ').partial('world').partial('🙂')()).toBe('hello world! 🙂')
+    })
+  })
+
+  describe('currying variadic functions', function () {
+    let sumMethod
+
+    beforeEach(function () {
```

Comment:
[This too](https://github.com/sithmel/iter-tools/pull/174/commits/a70990e282568420cfc70bbcfc98992e36974a7a#r259602797).

You are reassigning one `sumMethod` variable each time a new test run, this would be a problem when the tests are run in parallel.

Reply:
Jest never parallelizes inside a file. It will run different test files in parallel, and different projects as well, but each file is given it's own fresh execution environment and the tests inside are always run synchronously. Even if the tests are async, inside a single file each will wait for the previous to finish.

Reply:
Concurrency is not the only reason to prefer [my style](https://github.com/sithmel/iter-tools/pull/174/commits/a70990e282568420cfc70bbcfc98992e36974a7a#r259602797), there're also several others:
  * More concise and straightforward
  * Allow VSCode's type hint to help
  * Should function creation throw an error, the tests would fail immediately (I am too lazy to check whether errors in `beforeEach` prevents tests from being run)

Reply:
I do like it, actually. `beforeEach` will not run the test if it errors out, and it does work quite well when the setup can be expressed in a purely additive manner with nested describe blocks each adding a bit to the setup.

In simpler situations like this though it can be more useful to be able to pass a parameter to your setup, also there's the type hint, though I don't use vscode.

Reply:
Reading your response, I can't know for sure if you agree or disagree with me, so I still insist on this (but I guess it didn't matter since you're probably not going to merge this).

BTW, since you don't use VSCode, what editor do you use? (I'm just curious)

Reply:
I'm a Sublime text user.